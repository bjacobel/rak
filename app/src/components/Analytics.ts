import { useEffect, useMemo } from 'react';
import { useLocation, useSearch } from 'wouter';

import GA from '../services/GA';

export default () => {
  const location = useLocation();
  const search = useSearch();

  const ga = useMemo(() => new GA(), []);

  useEffect(() => {
    if (!ga) return;
    ga.set('page', location + search);
    ga.pageview();
  }, [location, search, ga]);

  return null;
};
