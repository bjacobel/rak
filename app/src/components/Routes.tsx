import { Switch, Route } from 'wouter';

import Main from './Main';
import Child from './Child';
import Error from './Error';
import NotFound from './NotFound';

export default () => (
  <Switch>
    <Route component={Main} path="/" />
    <Route component={Child} path="/child/:id" />
    <Route component={Error} path="/error" />
    <Route component={NotFound} path="*" />
  </Switch>
);
