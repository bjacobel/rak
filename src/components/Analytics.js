import React, { useEffect, useMemo } from 'react';
import { Router, useLocation } from '@reach/router';

import GA from '../services/GA';

const LocationTracker = () => {
  const location = useLocation();

  const ga = useMemo(() => new GA());

  useEffect(() => {
    if (!ga) return;
    ga.set('page', location.pathname + location.search);
    ga.pageview('send', 'pageview');
  }, [location, ga]);

  return null;
};

export default () => (
  <Router primary={false}>
    <LocationTracker path="*" />
  </Router>
);
