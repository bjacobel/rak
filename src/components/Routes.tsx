import React from 'react';
import { Routes, Route } from 'react-router-dom';

import NotFound from './NotFound';
import Child from './Child';
import Main from './Main';

export default () => (
  <Routes>
    <Route element={<Main />} path="/" />
    <Route element={<Child />} path="/child/:id" />
    <Route element={<NotFound />} path="*" />
  </Routes>
);
