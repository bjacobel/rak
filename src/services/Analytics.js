/* eslint-disable class-methods-use-this */

import { GA_ID, TRACK_ANALYTICS } from '../constants';

// @TODO: I'd really like to not install ga onto window; it should be self-contained by this class

export default class Analytics {
  constructor() {
    if (!window.ga) {
      window.ga =
        window.ga ||
        // eslint-disable-next-line func-names
        function(...args) {
          (window.ga.q = window.ga.q || []).push(args);
        };
      window.ga.l = +new Date();
      window.ga('create', GA_ID, 'auto');

      const script = document.createElement('script');
      script.async = true;
      script.src = 'https://www.google-analytics.com/analytics.js';
      document.body.appendChild(script);
    }
  }

  pageview() {
    if (TRACK_ANALYTICS) {
      window.ga('send', 'pageview');
    }
  }

  event(category, action, value) {
    if (TRACK_ANALYTICS) {
      window.ga('send', 'event', category, action, Number.isNaN(value) ? undefined : value);
    }
  }
}
