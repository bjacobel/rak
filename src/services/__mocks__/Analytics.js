const Analytics = jest.fn();

Analytics.prototype = {
  pageview: jest.fn(),
};

export default Analytics;
