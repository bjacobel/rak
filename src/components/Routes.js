import React, { Component } from 'react';
import { BrowserRouter, Route, Switch } from 'react-router-dom';

import NotFound from './NotFound';
import Child from './Child';
import Main from './Main';
import Analytics from '../services/Analytics';

class GARoute extends Route {
  render() {
    this.props.ga.pageview();
    // eslint-disable-next-line react/jsx-props-no-spreading
    return <Route {...this.props} />;
  }
}

export default class Routes extends Component {
  constructor() {
    super();
    this.ga = new Analytics();
  }

  render() {
    return (
      <>
        <BrowserRouter>
          <div>
            <Switch>
              <GARoute ga={this.ga} path="/" exact component={Main} />
              <GARoute ga={this.ga} path="/child/:id" component={Child} />
              <GARoute ga={this.ga} component={NotFound} />
            </Switch>
          </div>
        </BrowserRouter>
      </>
    );
  }
}
