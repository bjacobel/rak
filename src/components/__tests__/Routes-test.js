import React from 'react';
import { shallow, mount } from 'enzyme';

import Routes from 'components/Routes';
import Analytics from 'services/Analytics';

jest.mock('components/Main');
jest.mock('components/Child');
jest.mock('components/NotFound');
jest.mock('services/Analytics');

const setPath = value => {
  global.history.pushState({}, value, value);
};

describe('Router', () => {
  it('matches snapshot', () => {
    expect(shallow(<Routes />)).toMatchSnapshot();
  });

  describe('Routing config', () => {
    it('has a home route', () => {
      setPath('/');
      const routes = mount(<Routes />);
      expect(routes.find('Connect(Main)').length).toBe(1);
    });

    it('has a route for child views that takes a url param', () => {
      setPath('/child/1');
      const routes = mount(<Routes />);
      expect(routes.find('Child').length).toBe(1);
      expect(routes.find('Child').props()).toEqual(
        expect.objectContaining({
          match: expect.objectContaining({
            params: {
              id: '1',
            },
          }),
        }),
      );
    });

    it('has a fallthrough 404', () => {
      setPath('/asdfasaddf');
      const routes = mount(<Routes />);
      expect(routes.find('NotFound').length).toBe(1);
    });
  });

  describe('Analytics functionality', () => {
    beforeEach(() => {
      // reset call count before each test
      Analytics.prototype.pageview.mockClear();
    });

    it('inits the analytics class', () => {
      mount(<Routes />);
      expect(Analytics).toHaveBeenCalled();
    });

    it('calls the analytics pageview fn when you navigate to a new route', () => {
      mount(<Routes />);
      setPath('/child/1');
      expect(Analytics.prototype.pageview).toHaveBeenCalledTimes(1);
    });

    it('supplies GA to routed components via props', () => {
      const routes = mount(<Routes />);
      expect(routes.find('Route').props()).toEqual(
        expect.objectContaining({
          ga: {},
        }),
      );
    });
  });
});
