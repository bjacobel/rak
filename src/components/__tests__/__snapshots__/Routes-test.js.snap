// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router matches snapshot 1`] = `
<Fragment>
  <BrowserRouter>
    <div>
      <Switch>
        <GARoute
          component={
            Object {
              "$$typeof": Symbol(react.memo),
              "WrappedComponent": [MockFunction],
              "apply": [MockFunction],
              "bind": [MockFunction],
              "call": [MockFunction],
              "compare": null,
              "displayName": "Connect(Main)",
              "type": [MockFunction],
            }
          }
          exact={true}
          ga={Object {}}
          path="/"
        />
        <GARoute
          component={[MockFunction]}
          ga={Object {}}
          path="/child/:id"
        />
        <GARoute
          component={[MockFunction]}
          ga={Object {}}
        />
      </Switch>
    </div>
  </BrowserRouter>
</Fragment>
`;
